service/src/parentchain_block_syncer.rs

```diff
@@ -0,0 +1,154 @@
+		self.substrate_api.get_blocks(from, to)
+	}
+
+	fn sync_parentchain<PB: Block>(
+		&self,
+		blocks: &[SignedBlockT<PB>],
+		nonce: u32,
+	) -> EnclaveResult<()> {
+		self.enclave_api.sync_parentchain(blocks, nonce)
+	}
+}
+
+pub struct NodeApiDelegate<EnclaveApi: EnclaveBase + Sidechain> {
+	substrate_api: Api<sr25519::Pair, WsRpcClient>,
+	enclave_api: EnclaveApi,
+}
+
+impl<EnclaveApi: EnclaveBase + Sidechain> NodeApiDelegate<EnclaveApi> {
+	pub(crate) fn new(
+		substrate_api: Api<sr25519::Pair, WsRpcClient>,
+		enclave_api: EnclaveApi,
+	) -> Self {
+		NodeApiDelegate { substrate_api, enclave_api }
+	}
+}
+
+pub(crate) struct ParentchainBlockSyncer<NodeApi: NodeApiFacade> {
+	node_api: NodeApi,
+}
+
+impl<NodeApi> ParentchainBlockSyncer<NodeApi>
+where
+	NodeApi: NodeApiFacade,
+{
+	pub fn new(node_api: NodeApi) -> Self {
+		ParentchainBlockSyncer { node_api }
+	}
+
+	/// Fetches the amount of blocks to sync from the parentchain and feeds them to the enclave.
+	/// Returns the latest synced block Header.
+	///
+	pub fn sync_parentchain(&self, last_synced_header: Header) -> Header {
+		trace!("Getting current head");
+		let curr_block: SignedBlock = self.node_api.get_last_finalized_block();
+		let curr_block_number = curr_block.block.header.number;
+
+		let mut until_synced_header = last_synced_header;
+		loop {
+			let block_chunk_to_sync =
+				self.get_block_chunk_to_sync(until_synced_header.number, curr_block_number);
+			println!("[+] Found {} block(s) to sync", block_chunk_to_sync.len());
+			if block_chunk_to_sync.is_empty() {
+				return until_synced_header
+			}
+			let tee_nonce = self.node_api.tee_nonce();
+
+			if let Err(e) =
+				self.node_api.sync_parentchain(block_chunk_to_sync.as_slice(), tee_nonce)
+			{
+				error!("{:?}", e);
+				// enclave might not have synced
+				return until_synced_header
+			};
+			until_synced_header = block_chunk_to_sync
+				.last()
+				.map(|b| b.block.header.clone())
+				.expect("Chunk can't be empty; qed");
+			println!(
+				"Synced {} out of {} finalized parentchain blocks",
+				until_synced_header.number, curr_block_number,
+			)
+		}
+	}
+
+	/// gets a list of blocks that need to be synced, ordered from oldest to most recent header
+	/// starting from the last synced head until the current head but at most BLOCK_SYNC_BATCH_SIZE.
+	fn get_block_chunk_to_sync(
+		&self,
+		last_synced_head_number: BlockNumber,
+		curr_block_number: BlockNumber,
+	) -> Vec<SignedBlock> {
+		let no_blocks_to_sync = curr_block_number - last_synced_head_number;
+		if no_blocks_to_sync > 1 {
+			println!("light client is synced until block: {:?}", last_synced_head_number);
+			println!("Last finalized block number: {:?}\n", curr_block_number);
+		}
+		if no_blocks_to_sync > BLOCK_SYNC_BATCH_SIZE {
+			println!(
+				"Remaining blocks to fetch until last synced header: {:?}",
+				curr_block_number - last_synced_head_number
+			);
+		}
+
```

Comment:
I am giving the above suggestions because I think these prints are not necessary. We spam too much in the sync process, it has been bugging me for a long time.